src/modules/json.c

```diff
@@ -1,35 +1,9 @@
 #include "json.h"
-#include "utf8.h"
 #include "pdjson.h"
 
-// Extracted from DOME engine
-#define VM_ABORT(vm, error) do {\
-  wrenSetSlotString(vm, 0, error);\
-  wrenAbortFiber(vm, 0); \
-} while(false);
-
-#define ASSERT_SLOT_TYPE(vm, slot, type, fieldName) \
-  if (wrenGetSlotType(vm, slot) != type) { \
-    VM_ABORT(vm, #fieldName " was not " #type); \
-    return; \
-  }
-
-// Json API
-
-// We have to use C functions for escaping chars
-// because a bug in compiler throws error when using \ in strings
-// inside Wren files.
-// TODO: Check this in the future.
-enum JsonOptions {
-    JSON_OPTS_NIL = 0,
-    JSON_OPTS_ESCAPE_SLASHES = 1,
-    JSON_OPTS_ABORT_ON_ERROR = 2
-};
-
 json_stream jsonStream[1];
 
 void jsonStreamBegin(WrenVM * vm) {
-  ASSERT_SLOT_TYPE(vm, 1, WREN_TYPE_STRING, "value");
```

Comment:
Should we make sure we have a string on the Wren side then?  All the CLI APIs do this... so usually you have a wrapper function that does the validation then calls the foreign function, should be easy to find an example.

Reply:
Yes type validation can be done in Wren side of things. Unless C is desired instead.

Reply:
No, the tradition is do it on the Wren side... since this is the only/key place that we're handing data to C I think that's the key place we need to validate types on the Wren side... I assume that's why the CLI does all this validation before dumping things into C... I assume the VM handles wrong types much cleaner than C does... therefore this is more critical to get right with foreign functions than the entire API - which I'd honestly rather let the VM error messages do their own work.

So we'd have:

- `stream_begin(value)` API wrapper, does validation of string argument
- `foreign stream_begin_(value)` C method

Reply:
I prefer `foreign f_stream_begin(value)` C method :)

Reply:
Wait, are we talking about wren naming or C side naming? Justification please?

Reply:
Wren side.
just a convention to name foreign methods with `f_` so its clear we are dealing with C code.

Reply:
Is this idea only for wrappers?  I don't see how this would work with unwrapped functions:

```
class Time {
    foreign static now()
    foreign static highResolution()
}
```

?

Reply:
I think would only be valid for foreign functions that accepts params. Since they would need param validation before C calls

Reply:
I'm not 100% sold yet, but what the heck - sure go for it.  I do see what you're trying to do with it, and that makes sense I think. :-)

Reply:
sent